### ISSUE TITLE
Improve the way we handle Enchatments (Attach Enchantment to target Object)

### ISSUE THREAD

**Issue by sIKE23** [sIKE23/Mage-Wars#11]
Automate the casting cost of hidden enchantments by attaching them to the target card and paying the 2 mana cost.

**Comment by anandaguneratne** [comment]
So, now that I can test multiplayer, I have run into a bizarre bug that prevents the alignment code from working properly when called using remoteCall(). It will take me a bit longer to figure out how to get around this.

**Comment by anandaguneratne** [comment]
Let me know when you plan to issue the update. I am almost done with attachCards (just 2 obnoxious bugs to iron out), and will try to upload it as soon as it is in decent shape. (also, I don't want my data to be accidentally overwritten by the update.

**Comment by anandaguneratne** [comment]
Okay, I just had a major breakthrough on the most critical bug, so the rest of the work I need to do is just fixing little quirks (which shouldn't take very long). I guess what I am saying is: hold off for a bit on the update, because I can make it a lot more interesting.

**Comment by anandaguneratne** [comment]
All right, the attachment code is good to go. The last significant remaining issue with it is that selecting and moving a stack of cards causes weird behavior; I have opened an issue on this problem on the main OCTGN issues page (https://github.com/kellyelton/OCTGN/issues/1348).

The main characteristics to note are:
-----
Features:
- The attachments of each card are preserved in the order in which they were attached to that card.
- Removing a card from the table detaches everything that was attached to it.
- Two methods of attaching a card; explicitly, using targeting and alt+q, and by simply dragging one
card over the other (if autoAttach is enabled), using the moveToCard method in actions.py
- Detaching a card is as easy as dragging it away from the card to which it was attached

Known Bugs/Missing Features:
- Mass selecting a stack of cards and dragging it causes the cards to unattach and semi-randomly attach
to each other
- If a card leaves the table, the attachments are automatically sent to the middle of the board. Not a
problem in itself, but perhaps a better location could be found...
- Right now, the criteria for what may be attached to what are pretty loose. It will be easy to add
restrictions, however.

I think I have (sort of) figured out how github works; right now, I am re-cloning 'Mage-Wars' right now, and once that is done I will commit the three files (action.py, attachCard.py, and definitions.xml) containing the changes.

**Comment by anandaguneratne** [comment]
And it's committed! Let me know what you think.

**Comment by sIKE23** [comment]
To what branch?

**Comment by anandaguneratne** [comment]
Ok, I think I made a mistake the first time I committed it. Might have sent it to the master branch. I think I have fixed that.

**Comment by anandaguneratne** [comment]
Do I need to click 'sync'?

**Comment by sIKE23** [comment]
Yes....that commits it to the repository....

**Comment by anandaguneratne** [comment]
Okay, I think I got it. Do you see the changes?

**Comment by sIKE23** [comment]
Yes I see the changes, I am trying to figure out what happened to
actions.py. There are a lot of changes.....

//FC

**Comment by anandaguneratne** [comment]
Looking at the changes, I just realized that I was not working from the file that was in the attachCards branch, but from some other file (v2.8.1.3). So most of the changes in actions.py are irrelevant.

**Comment by sIKE23** [comment]
Ok, that makes a bit more sense....gonna have to dig through and figure out
what needs to be done...

**Comment by anandaguneratne** [comment]
I'm going to revert the changes and insert the relevant ones back in.

**Comment by anandaguneratne** [comment]
Okay, I just committed changes based off of the correct version. I'm going to test them out, to be sure, but it should be clearer what has been done now.

**Comment by anandaguneratne** [comment]
Well, after that last commit, some aspects of the attachCards module are no longer working as intended. I'm not sure why. I'm going to have to take another look at the differences.

**Comment by sIKE23** [comment]
ACG,

Ok, that is always the fun part of GitHub that I try to shield you from as
much as I can. If we want to merge the work from the Master branch in that
of the AttachCards branch we will have even more fun.

//FC

**Comment by sIKE23** [comment]
I pulled your bits out and put it into the actions.py that I have been
working with. The attachment stuff works for me.

**Comment by anandaguneratne** [comment]
Okay, I figured out the reason I was having trouble reproducing the earlier results after your update, and it is pretty silly; after finalizing the code the first time around, I immediately switched my OCTGN filepath back to my non-dev directory (I was afraid an update might wipe my directory), and then promptly forgot that I had done that. So, as you say, the attachment stuff all works! (apart from the multiple selection movement issue)

**Comment by anandaguneratne** [comment]
Thinking about how the code for this works, I suspect there might be a multiplayer bug in which card alignment fails for 3 or more players under certain arrangements of cards. Once I have a chance, I'll test this and report back.

**Comment by anandaguneratne** [comment]
Just tested it. As I thought, in a 3 player game the alignment function will occasionally misorder the stack of attached cards. This is not a huge bug, (won't crash or break the game), and most players will not even notice it since multiplayer games are relatively rare. Still, it will need to be addressed at some point. I am not sure how to fix it at the moment, but I'll think on it (and create a separate issue for it).

**Comment by anandaguneratne** [comment]
The behavior when removing a card from the board is greatly improved with https://github.com/sIKE23/Mage-Wars/commit/6127b05f0fe582b5d8c66cef2a4e18d29ed6ec8f. Previously, it simply stacked all attachments in the center of the board; now, it separates them according to controller and stacks them neatly in front of their controllers (just above the place where double-clicking will autoplace a card).

**Comment by anandaguneratne** [comment]
https://github.com/sIKE23/Mage-Wars/commit/c82d09fd6a214817e49feb42a14c4dc9219c6742 fixes the master branch so that attachments work. Now I just need to do the same for Q1 -- 2015.

**Comment by sIKE23** [comment]
Done...

**Comment by anandaguneratne** [comment]
Great! Everything seems to be working.

**Comment by anandaguneratne** [comment]
Oops! I accidentally clicked close and comment.

**Comment by sIKE23** [comment]
This is not working in the Master Branch

**Comment by sIKE23** [comment]
Never mind

**Comment by sIKE23** [comment]
Send to Front/Send to Back (Page Up/Page Down) have some weird and unintended interactions with card attachments. Nothing we have to solve for this release, but we an issue we found.

**Comment by sIKE23** [comment]
I am happy with this for this release, closing out! Great work!
